### PR TITLE
Support deserialization of V3 StoreFindToken to incorporate reset key

### PR DESCRIPTION
**Pull request by jsjtzyy**
Introduced reset key into StoreFindToken with aim to help find closest valid index segment for replication token. This is useful when replication token is invalidated by source node. Instead of resetting it to very beginning, reset key may help find valid starting point without losing too much ground.

### CODE REVIEW

File: ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java [hunk 1]
```diff
@@ -105,18 +116,23 @@
    * @param inclusive {@code true} if the offset is inclusive or in other words the blob at the given offset is inclusive.
    *                  {@code false} otherwise
    * @param version refers to the version of the token
+   * @param resetKey The {@link StoreKey} used to find closest valid starting point in case the index segment the token
+   *                 refers to has been compacted.
    */
-  private StoreFindToken(FindTokenType type, Offset offset, StoreKey key, UUID sessionId, UUID incarnationId,
-      boolean inclusive, short version) {
+   StoreFindToken(FindTokenType type, Offset offset, StoreKey key, UUID sessionId, UUID incarnationId,
+      boolean inclusive, short version, StoreKey resetKey) {
```
Comment (jsjtzyy) [inline ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java:123]: For more context, feel free to take a look at the diagram in this doc: https://docs.google.com/document/d/19m1MNMH4ITNZYVXfCUBzYEc5wBl0IpI7i0uTGVRS8SI/edit
and more details in original doc:
https://docs.google.com/document/d/1Lg2Alp5jPfD9NkvDExi0LuLKuurRllV6l3s2-D5YCLg/edit

**Review by justinlin-linkedin — APPROVED** [review]
LGTM

File: ambry-store/src/test/java/com/github/ambry/store/StoreFindTokenTest.java [hunk 2]
```diff
@@ -273,35 +347,109 @@ static DataInputStream getSerializedStream(StoreFindToken token, short version)
...
-        Utils.serializeNullableString(bufWrap, sessionId == null ? null : sessionId.toString());
+        bufWrap.putInt(sessionIdBytes.length);
+        bufWrap.put(sessionIdBytes);
         // add type
-        type = token.getType();
         bufWrap.putShort((byte) type.ordinal());
         bufWrap.put(offsetBytes);
-        if (key != null) {
-          bufWrap.put(key.toBytes());
+        if (storeKeyInBytes.length > 0) {
```
Comment (justinlin-linkedin) [inline ambry-store/src/test/java/com/github/ambry/store/StoreFindTokenTest.java:370]: how is this method different from `StoreFindToken.toBytes`

File: ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java [hunk 3]
```diff
@@ -312,10 +433,10 @@ public long getBytesRead() {
...
         break;
       case VERSION_2:
-        offsetBytes = offset != null ? offset.toBytes() : ZERO_LENGTH_ARRAY;
-        sessionIdBytes = sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
-        byte[] incarnationIdBytes = incarnationId != null ? incarnationId.toString().getBytes() : ZERO_LENGTH_ARRAY;
-        storeKeyBytes = storeKey != null ? storeKey.toBytes() : ZERO_LENGTH_ARRAY;
+        offsetBytes = getOffsetInBytes();
+        sessionIdBytes = getSessionIdInBytes();
+        byte[] incarnationIdBytes = getIncarnationIdInBytes();
+        storeKeyBytes = getStoreKeyInBytes();
```
[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java:439]: This PR supports deserialization only. Once it is deployed to everywhere, I will add serialization in a separate PR.

File: ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java [hunk 4]
```diff
@@ -353,6 +474,22 @@ public long getBytesRead() {
...
+
+  byte[] getStoreKeyInBytes() {
+    return storeKey != null ? storeKey.toBytes() : ZERO_LENGTH_ARRAY;
+  }
+
+  byte[] getSessionIdInBytes() {
+    return sessionId != null ? sessionId.toString().getBytes() : ZERO_LENGTH_ARRAY;
+  }
+
+  byte [] getIncarnationIdInBytes() {
```
[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java:489]: sure, will do

File: ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java [hunk 5]
```diff
@@ -389,13 +532,16 @@ public boolean equals(Object o) {
...
       return false;
     }
     if (inclusive != that.inclusive) {
       return false;
     }
-    return storeKey != null ? storeKey.equals(that.storeKey) : that.storeKey == null;
+    if (!Objects.equals(storeKey, that.storeKey)) {
+      return false;
+    }
+    return Objects.equals(resetKey, that.resetKey) && Objects.equals(resetKeyType, that.resetKeyType);
```
[the remark this answers is not in the document]
Comment (jsjtzyy) [reply ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java:544]: Yeah, make sense.

File: ambry-store/src/test/java/com/github/ambry/store/StoreFindTokenTest.java (hunk 2 above)
Comment (jsjtzyy) [reply ambry-store/src/test/java/com/github/ambry/store/StoreFindTokenTest.java:370]: This was written to test deserialization of new token before production code supports serializing new token.

**Review by justinlin-linkedin — APPROVED** [review]

File: ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java [hunk 6]
```diff
@@ -69,30 +78,73 @@
...
+   * VERSION_3
+   * +--------------------------------------------------------------------------------------------------------------------------+
+   * | version | token type | incarnation id | session id | index segment start offset | store key | reset key | reset key type |
+   * | (short) |  (short)   |  (n bytes)     |  (n bytes) |        (n bytes)           | (n bytes) | (n bytes) |    (short)     |
+   * +--------------------------------------------------------------------------------------------------------------------------+
    * @param key The {@link StoreKey} which the token refers to. Index segments are keyed on store keys and hence
    * @param indexSegmentStartOffset the start offset of the index segment which the token refers to
    * @param sessionId the sessionId of the store
    * @param incarnationId the incarnationId of the store
+   * @param resetKey The {@link StoreKey} used to find closest valid starting point in case the index segment the token
```
Comment (ankagrawal) [inline ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java:108]: Can you please rephrase this comment. I am not sure what this means.

File: ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java [hunk 7]
```diff
@@ -69,30 +78,73 @@
...
+   * +----------------------------------------------------------------------------------------------------------+
+   * | version | token type | incarnation id | session id | log offset | inclusive | reset key | reset key type |
+   * | (short) |  (short)   |   (n bytes)    | (n bytes)  | (n bytes)  |  (byte)   | (n bytes) |    (short)     |
+   * +----------------------------------------------------------------------------------------------------------+
    * @param offset the offset that this token refers to in the journal
    * @param sessionId the sessionId of the store
    * @param incarnationId the incarnationId of the store
    * @param inclusive {@code true} if the offset is inclusive or in other words the blob at the given offset is inclusive.
    *                  {@code false} otherwise
+   * @param resetKey The {@link StoreKey} used to find closest valid starting point in case the index segment the token
```
Comment (ankagrawal) [inline ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java:141]: same as above.

File: ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java [hunk 8]
```diff
@@ -105,18 +157,23 @@
    * @param inclusive {@code true} if the offset is inclusive or in other words the blob at the given offset is inclusive.
    *                  {@code false} otherwise
    * @param version refers to the version of the token
+   * @param resetKey The {@link StoreKey} used to find closest valid starting point in case the index segment the token
```
Comment (ankagrawal) [inline ambry-store/src/main/java/com/github/ambry/store/StoreFindToken.java:160]: same as above.

**Review by ankagrawal — APPROVED** [review]
Looks good to me after addressing the comments.

**Review by ankagrawal — APPROVED** [review]
lgtm

**Pull request merged**